PR: https://github.com/odoo/odoo/pull/222359

From: 6d7f4012cfa06b35089e7557522fcbf658978b37
From: Natalie (naja)
Date: 2025-09-09 12:52:58

Breaking data model changes scores: del:2.4 + add:0, change matches:
-    overtime_id = fields.Many2one('hr.attendance.overtime', string='Extra Hours')
-    employee_overtime = fields.Float(related='employee_id.total_overtime', groups='base.group_user')
+    employee_overtime = fields.Float(compute='_compute_employee_overtime', groups='base.group_user')
-    overtime_id = fields.Many2one('hr.attendance.overtime', string='Extra Hours', groups='hr_holidays.group_hr_holidays_user')

Total Changes: 216

[IMP] hr_attendance: Add overtime rulesets

task-4577691

closes odoo/odoo#222359

Related: odoo/enterprise#93528
Related: odoo/upgrade#8394
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_holidays_attendance/models/__init__.py
+++ b/addons/hr_holidays_attendance/models/__init__.py
@@ -1,7 +1,10 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
+from . import hr_attendance_overtime_rule
+from . import hr_attendance_overtime
 from . import hr_leave_allocation
 from . import hr_leave_accrual_plan_level
 from . import hr_leave_type
 from . import hr_leave
 from . import ir_ui_menu
+from . import hr_employee

--- a/None
+++ b/addons/hr_holidays_attendance/models/hr_attendance_overtime.py
@@ -0,0 +1,10 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, fields
+
+
+class HrAttendanceOvertimeLine(models.Model):
+    _name = 'hr.attendance.overtime.line'
+    _inherit = 'hr.attendance.overtime.line'
+
+    compensable_as_leave = fields.Boolean("Compensable as Time Off")

--- a/None
+++ b/addons/hr_holidays_attendance/models/hr_attendance_overtime_rule.py
@@ -0,0 +1,31 @@
+
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, fields
+
+
+class HrAttendanceOvertimeRule(models.Model):
+    _name = 'hr.attendance.overtime.rule'
+    _inherit = 'hr.attendance.overtime.rule'
+
+    compensable_as_leave = fields.Boolean("Give back as time off", default=False)
+
+    def _extra_overtime_vals(self):
+        if not self:
+            return {
+                **super()._extra_overtime_vals(),
+                'compensable_as_leave': False,
+            }
+
+        res = super()._extra_overtime_vals()
+        res['compensable_as_leave'] = any(self.mapped('compensable_as_leave'))
+        if self.ruleset_id.rate_combination_mode == 'sum' and any(self.mapped('paid')):
+            combined_rate = 1.0
+            combined_rate += sum(r.amount_rate - 1.0 for r in self.filtered(
+                lambda r: r.paid and not r.compensable_as_leave
+            ))
+            combined_rate += sum(r.amount_rate for r in self.filtered(
+                lambda r: r.paid and r.compensable_as_leave
+            ))
+            res['amount_rate'] = combined_rate
+        return res

--- a/None
+++ b/addons/hr_holidays_attendance/models/hr_employee.py
@@ -0,0 +1,30 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import models, api
+from odoo.fields import Domain
+
+
+class HrEmployee(models.Model):
+    _inherit = "hr.employee"
+
+    @api.model
+    def get_overtime_data(self, domain=None, employee_id=None):
+        overtime_data = super().get_overtime_data(domain, employee_id)
+        domain = [] if domain is None else domain
+        overtime_adjustments = {
+            allocation[0].id: allocation[1]
+            for allocation in self.env["hr.leave.allocation"]._read_group(
+                domain=Domain.AND(
+                    domain,
+                    [
+                        ('holiday_status_id.overtime_deductible', '=', True),
+                        ('state', '=', 'confirm'),
+                    ]),
+                groupby=['employee_id'],
+                aggregates=['number_of_hours_display:sum']
+            )
+        }
+        overtime_data.update({
+            'overtime_adjustments': overtime_adjustments
+        })
+        return overtime_data

--- a/addons/hr_holidays_attendance/models/hr_leave.py
+++ b/addons/hr_holidays_attendance/models/hr_leave.py
@@ -1,4 +1,3 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from collections import defaultdict
@@ -12,8 +11,7 @@ from odoo.tools import float_round
 class HrLeave(models.Model):
     _inherit = 'hr.leave'
 
-    overtime_id = fields.Many2one('hr.attendance.overtime', string='Extra Hours')
-    employee_overtime = fields.Float(related='employee_id.total_overtime', groups='base.group_user')
+    employee_overtime = fields.Float(compute='_compute_employee_overtime', groups='base.group_user')
     overtime_deductible = fields.Boolean(compute='_compute_overtime_deductible')
 
     @api.depends('holiday_status_id')
@@ -32,41 +30,60 @@ class HrLeave(models.Model):
         fields_to_check = {'number_of_days', 'request_date_from', 'request_date_to', 'state', 'employee_id', 'holiday_status_id'}
         if not any(field for field in fields_to_check if field in vals):
             return res
-        if vals.get('holiday_status_id'):
-            self._check_overtime_deductible(self)
-        #User may not have access to overtime_id field
-        for leave in self.sudo().filtered('overtime_id'):
-            # It must always be possible to refuse leave based on overtime
-            if vals.get('state') in ['refuse']:
-                continue
-            employee = leave.employee_id
-            duration = leave.number_of_hours
-            overtime_duration = leave.overtime_id.sudo().duration
-            if overtime_duration != -1 * duration:
-                if duration > employee.total_overtime - overtime_duration:
-                    raise ValidationError(_('The employee does not have enough extra hours to extend this leave.'))
-                leave.overtime_id.sudo().duration = -1 * duration
+        self._check_overtime_deductible(self)
         return res
 
+    @api.model
+    def _get_deductible_employee_overtime(self, employees):
+        # return dict {employee: number of hours}
+        diff_by_employee = defaultdict(lambda: 0)
+        for employee, hours in self.env['hr.attendance.overtime.line'].sudo()._read_group(
+            domain=[
+                ('compensable_as_leave', '=', True),
+                ('employee_id', 'in', employees.ids),
+                ('status', '=', 'approved'),
+            ],
+            groupby=['employee_id'],
+            aggregates=['manual_duration:sum'],
+        ):
+            diff_by_employee[employee] += hours
+        for employee, hours in self._read_group(
+            domain=[
+                ('holiday_status_id.overtime_deductible', '=', True),
+                ('holiday_status_id.requires_allocation', '=', False),
+                ('employee_id', 'in', employees.ids),
+                ('state', 'not in', ['refuse', 'cancel']),
+            ],
+            groupby=['employee_id'],
+            aggregates=['number_of_hours:sum'],
+        ):
+            diff_by_employee[employee] -= hours
+        for employee, hours in self.env['hr.leave.allocation']._read_group(
+            domain=[
+                ('holiday_status_id.overtime_deductible', '=', True),
+                ('employee_id', 'in', employees.ids),
+                ('state', '=', 'confirm'),
+            ],
+            groupby=['employee_id'],
+            aggregates=['number_of_hours_display:sum'],
+        ):
+            diff_by_employee[employee] -= hours
+        return diff_by_employee
+
+    @api.depends('number_of_hours', 'employee_id', 'holiday_status_id')
+    def _compute_employee_overtime(self):
+        diff_by_employee = self._get_deductible_employee_overtime(self.employee_id)
+        for leave in self:
+            leave.employee_overtime = diff_by_employee[leave.employee_id]
+
     def _check_overtime_deductible(self, leaves):
         # If the type of leave is overtime deductible, we have to check that the employee has enough extra hours
-        for leave in leaves:
-            if not leave.overtime_deductible:
-                leave.sudo().overtime_id.unlink()
-                continue
-            employee = leave.employee_id.sudo()
-            duration = leave.number_of_hours
-            if not leave.sudo().overtime_id:
-                if duration > employee.total_overtime:
-                    if employee.user_id == self.env.user:
-                        raise ValidationError(_('You do not have enough extra hours to request this leave'))
-                    raise ValidationError(_('The employee does not have enough extra hours to request this leave.'))
-                leave.sudo().overtime_id = self.env['hr.attendance.overtime'].sudo().create({
-                    'employee_id': employee.id,
-                    'date': leave.date_from,
-                    'adjustment': True,
-                    'duration': -1 * duration,
-                })
+        hours = self._get_deductible_employee_overtime(leaves.employee_id)
+        for leave in leaves.filtered('overtime_deductible'):
+            if hours[leave.employee_id] < 0:
+                if leave.employee_id.user_id == self.env.user:
+                    raise ValidationError(_('You do not have enough extra hours to request this leave'))
+                raise ValidationError(_('The employee does not have enough extra hours to request this leave.'))
 
     def action_reset_confirm(self):
         self._check_overtime_deductible(self)
@@ -80,7 +97,6 @@ class HrLeave(models.Model):
 
     def action_refuse(self):
         res = super().action_refuse()
-        self.sudo().overtime_id.unlink()
         return res
 
     def _validate_leave_request(self):
@@ -93,20 +109,17 @@ class HrLeave(models.Model):
         return res
 
     def _update_leaves_overtime(self):
-        employee_dates = defaultdict(set)
-        for leave in self:
-            if leave.employee_id:
-                for d in range((leave.date_to - leave.date_from).days + 1):
-                    # Sudo employee to get info from version.
-                    employee_dates[leave.employee_id].add(self.env['hr.attendance']._get_day_start_and_day(leave.employee_id.sudo(), leave.date_from + timedelta(days=d)))
-        if employee_dates:
-            self.env['hr.attendance'].sudo()._update_overtime(employee_dates)
-
-    def unlink(self):
-        # TODO master change to ondelete
-        self.sudo().overtime_id.unlink()
-        return super().unlink()
+        Attendance = self.env['hr.attendance']
+        dates = [
+            Attendance._attendance_date(leave.date_from, leave.employee_id)
+            for leave in self.filtered(lambda leave: leave.state == 'confirmed')
+        ]
+        if dates:
+            Attendance.search([
+                ('date', '>=', min(dates)),
+                ('date', '<=', max(dates)),
+                ('employee_id', 'in', self.employee_id.ids),
+            ])._update_overtimes()
 
     def _force_cancel(self, *args, **kwargs):
         super()._force_cancel(*args, **kwargs)
-        self.sudo().overtime_id.unlink()

--- a/addons/hr_holidays_attendance/models/hr_leave_allocation.py
+++ b/addons/hr_holidays_attendance/models/hr_leave_allocation.py
@@ -23,7 +23,6 @@ class HrLeaveAllocation(models.Model):
         return res
 
     overtime_deductible = fields.Boolean(compute='_compute_overtime_deductible')
-    overtime_id = fields.Many2one('hr.attendance.overtime', string='Extra Hours', groups='hr_holidays.group_hr_holidays_user')
     employee_overtime = fields.Float(related='employee_id.total_overtime', groups='base.group_user')
 
     @api.depends('holiday_status_id')
@@ -34,18 +33,11 @@ class HrLeaveAllocation(models.Model):
     @api.model_create_multi
     def create(self, vals_list):
         res = super().create(vals_list)
+        deductible = self.env['hr.leave']._get_deductible_employee_overtime(res.employee_id)
         for allocation in res:
             if allocation.overtime_deductible:
-                duration = allocation.number_of_hours_display
-                if duration > allocation.employee_id.sudo().total_overtime:
+                if deductible[allocation.employee_id] < 0:
                     raise ValidationError(_('The employee does not have enough overtime hours to request this leave.'))
-                if not allocation.overtime_id:
-                    allocation.sudo().overtime_id = self.env['hr.attendance.overtime'].sudo().create({
-                        'employee_id': allocation.employee_id.id,
-                        'date': allocation.date_from,
-                        'adjustment': True,
-                        'duration': -1 * duration,
-                    })
         return res
 
     def write(self, vals):
@@ -54,19 +46,14 @@ class HrLeaveAllocation(models.Model):
             return res
         if not self.env.user.has_group("hr_holidays.group_hr_holidays_user") and any(allocation.state not in ('draft', 'confirm') for allocation in self):
             raise ValidationError(_('Only an Officer or Administrator is allowed to edit the allocation duration in this status.'))
-        for allocation in self.sudo().filtered('overtime_id'):
-            employee = allocation.employee_id
-            duration = allocation.number_of_hours_display
-            overtime_duration = allocation.overtime_id.sudo().duration
-            if overtime_duration != -1 * duration:
-                if duration > employee.total_overtime - overtime_duration:
-                    raise ValidationError(_('The employee does not have enough extra hours to extend this allocation.'))
-                allocation.overtime_id.sudo().duration = -1 * duration
+        deductible = self.env['hr.leave']._get_deductible_employee_overtime(self.employee_id)
+        for allocation in self.sudo().filtered('overtime_deductible'):
+            if deductible[allocation.employee_id] < 0:
+                raise ValidationError(_('The employee does not have enough overtime hours to request this leave.'))
         return res
 
     def action_refuse(self):
         res = super().action_refuse()
-        self.overtime_id.sudo().unlink()
         return res
 
     def _get_accrual_plan_level_work_entry_prorata(self, level, start_period, start_date, end_period, end_date):

--- a/addons/hr_holidays_attendance/models/hr_leave_type.py
+++ b/addons/hr_holidays_attendance/models/hr_leave_type.py
@@ -1,4 +1,3 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo.tools.misc import format_duration
@@ -20,7 +19,8 @@ class HrLeaveType(models.Model):
             return super()._compute_display_name()
 
         employee = self.env['hr.employee'].browse(self.env.context.get('employee_id')).sudo()
-        if employee.total_overtime <= 0:
+        unspent_overtime = self.env['hr.leave']._get_deductible_employee_overtime(employee)[employee]
+        if unspent_overtime <= 0:
             return super()._compute_display_name()
 
         overtime_leaves = self.filtered(lambda l_type: l_type.overtime_deductible and not l_type.requires_allocation)
@@ -28,7 +28,7 @@ class HrLeaveType(models.Model):
             leave_type.display_name = "%(name)s (%(count)s)" % {
                 'name': leave_type.name,
                 'count': _('%s hours available',
-                    format_duration(employee.total_overtime)),
+                    format_duration(unspent_overtime)),
             }
         super(HrLeaveType, self - overtime_leaves)._compute_display_name()
 
